src/librustc_parse/parser/stmt.rs

```diff
@@ -323,22 +326,83 @@ impl<'a> Parser<'a> {
         Ok((self.parse_inner_attributes()?, self.parse_block_tail(lo, blk_mode)?))
     }
 
+    fn parse_block_tail(&mut self, lo: Span, s: BlockCheckMode) -> PResult<'a, P<Block>> {
+        if self.token.is_ident() && self.look_ahead(1, |t| t == &token::Colon) {
+            // We might be having a struct literal where people forgot to include the path:
+            // fn foo() -> Foo {
+            //     field: value,
+            // }
+            let mut snapshot = self.clone();
+            let path = Path { segments: vec![], span: self.prev_token.span.shrink_to_lo() };
+            let struct_expr = snapshot.parse_struct_expr(path, AttrVec::new(), false);
+            let block_tail = self.parse_block_tail_inner(lo, s, false);
+            return match (struct_expr, block_tail) {
+                (Ok(expr), Err(mut err)) => {
+                    // We have encountered for a fact the following:
```

Comment:
```suggestion
                    // We have encountered the following:
```
nit: "for a fact" just made this sentence a little harder to parse.